src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,694 @@
+        }
+        return perftest_mad_sm_query(ca, ca_port, mad_port, guid, dst_port);
+    } else if (!strncmp(addr, lid_str, strlen(lid_str))) {
+        addr += strlen(lid_str);
+
+        lid = strtol(addr, NULL, 0);
+        if (!IB_LID_VALID(lid)) {
+            return UCS_ERR_INVALID_PARAM;
+        }
+        return ib_portid_set(dst_port, lid, 0, 0) ? UCS_ERR_NO_DEVICE : UCS_OK;
+    }
+
+    ucs_error("MAD: invalid dst address, use '%s' or '%s' prefix", guid_str,
+              lid_str);
+    return UCS_ERR_INVALID_PARAM;
+}
+
+static int perftest_mad_accept_is_valid(void *buf, size_t size)
+{
+    perftest_params_t *params = buf;
+    size_t array_size;
+
+    if (size < sizeof(*params)) {
+        return 0;
+    }
+
+    array_size  = sizeof(*params->super.msg_size_list);
+    array_size *= params->super.msg_size_cnt;
+
+    return size == (sizeof(*params) + array_size);
+}
+
+static ucs_status_t perftest_mad_accept(perftest_mad_rte_group_t *rte_group,
+                                        struct perftest_context *ctx)
+{
+    size_t size;
+    ucs_status_t status;
+    void *ptr;
+    uint8_t buf[4096];
+    int lid;
+
+    free(ctx->params.super.msg_size_list);
+    ctx->params.super.msg_size_list = NULL;
+
+    do {
+        size   = sizeof(buf);
+        status = perftest_mad_recv(rte_group, buf, &size, &rte_group->dst_port);
+
+        ucs_debug("MAD: Accept: receive got status:%d, size:%zu/%zu", status,
+                  size, sizeof(buf));
+    } while ((status != UCS_OK) || !perftest_mad_accept_is_valid(buf, size));
+
+    lid = rte_group->dst_port.lid;
+    ucs_debug("MAD: Accept: remote lid:%d/0x%02x", lid, lid);
+
+    /* Import received message size list */
+    size = sizeof(*ctx->params.super.msg_size_list) *
+           ctx->params.super.msg_size_cnt;
+    ptr  = calloc(1, size);
+    if (ptr == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+    memcpy(&ctx->params, buf, sizeof(ctx->params));
+    ctx->params.super.msg_size_list = ptr;
+
+    memcpy(ctx->params.super.msg_size_list, buf + sizeof(ctx->params), size);
+
+    return perftest_mad_send(rte_group, &mad_magic, sizeof(mad_magic));
+}
+
+static ucs_status_t perftest_mad_connect(perftest_mad_rte_group_t *rte_group,
+                                         struct perftest_context *ctx)
+{
+    int iovcnt = 1;
+    ucs_status_t status;
+    size_t size;
+    struct iovec iov[2];
+
+    iov[0].iov_base = &ctx->params;
+    iov[0].iov_len  = sizeof(ctx->params);
+
+    size  = sizeof(*ctx->params.super.msg_size_list);
+    size *= ctx->params.super.msg_size_cnt;
+
+    if (size != 0) {
+        iov[1].iov_base = ctx->params.super.msg_size_list;
+        iov[1].iov_len  = size;
+        iovcnt++;
+    }
+    status = perftest_mad_sendv(rte_group, iov, iovcnt);
+    if (status != UCS_OK) {
+        return status;
+    }
+    return perftest_mad_recv_magic(rte_group, mad_magic);
+}
+
+static void perftest_mad_set_logging(void)
+{
+#define IB_DEBUG_LEVEL 10
```

Comment:
why is it an ifdef? can we take it from some mad header? or use static const int?